x-pack/plugins/security/public/components/management/users/edit_user.js

```diff
@@ -0,0 +1,442 @@
+    try {
+      await apiClient.changePassword(user.username, password, currentPassword);
+      toastNotifications.addSuccess('Password changed.');
+    } catch (e) {
+      if (e.status === 401) {
+        return this.setState({ currentPasswordError: true });
+      } else {
+        toastNotifications.addDanger(`Error setting password: ${e.data.message}`);
+      }
+    }
+    this.clearPasswordForm();
+  };
+  saveUser = async () => {
+    const { apiClient, changeUrl } = this.props;
+    const { user, password, selectedRoles } = this.state;
+    const userToSave = { ...user };
+    userToSave.roles = selectedRoles.map(selectedRole => {
+      return selectedRole.label;
+    });
+    if (password) {
+      userToSave.password = password;
+    }
+    try {
+      await apiClient.saveUser(userToSave);
+      toastNotifications.addSuccess(`Saved user ${user.username}`);
+      changeUrl(USERS_PATH);
+    } catch (e) {
+      toastNotifications.addDanger(`Error saving user: ${e.data.message}`);
+    }
+  };
+  clearPasswordForm = () => {
+    this.setState({
+      showChangePasswordForm: false,
+      password: null,
+      confirmPassword: null,
+    });
+  };
+  passwordFields = () => {
+    const { user, currentUser } = this.state;
+    const userIsLoggedInUser = user.username && user.username === currentUser.username;
+    return (
+      <Fragment>
+        {userIsLoggedInUser ? (
+          <EuiFormRow
+            label="Current password"
+            isInvalid={!!this.currentPasswordError()}
+            error={this.currentPasswordError()}
+          >
+            <EuiFieldText
+              name="currentPassword"
+              type="password"
+              onChange={event => this.setState({ currentPassword: event.target.value })}
+            />
+          </EuiFormRow>
+        ) : null}
+        <EuiFormRow
+          label={userIsLoggedInUser ? "New password" : "Password"}
+          isInvalid={!!this.passwordError()}
+          error={this.passwordError()}
+        >
+          <EuiFieldText
+            data-test-subj="passwordInput"
+            name="password"
+            type="password"
+            onChange={event => this.setState({ password: event.target.value })}
+            onBlur={event => this.setState({ password: event.target.value || '' })}
+          />
+        </EuiFormRow>
+        <EuiFormRow
+          label="Confirm password"
+          isInvalid={!!this.confirmPasswordError()}
+          error={this.confirmPasswordError()}
+        >
+          <EuiFieldText
+            data-test-subj="passwordConfirmationInput"
+            onChange={event => this.setState({ confirmPassword: event.target.value })}
+            onBlur={event => this.setState({ confirmPassword: event.target.value || '' })}
+            name="confirm_password"
+            type="password"
+          />
+        </EuiFormRow>
+      </Fragment>
+    );
+  };
+  changePasswordForm = () => {
```

Comment:
I think if we extracted this into its own component it would bring a lot of its own logic with it out of this file, and make this file a bit easier to follow.

Reply:
Hmm, I thought about that, but since password fields appear in the new user form as well, would make it tricky and possibly introduce duplication of code.